### ISSUE TITLE
Why is it called Q?

### ISSUE THREAD

**Issue by hguiney** [kriskowal/q#722]
I know this isn't a bug per se, but I was looking for reasoning in the README and wiki and couldn't find any. Why isn't it called "P" for "Promise"? Or "D" for "Deferred"? Is it named after the Star Trek or James Bond characters? Is it arbitrary? I'm confused enough by promises as it is, and the name "Q" just makes it seem even more cryptic and intimidating to me. I initially assumed it was a variable name in some sort of mathematical formula. The logo seems to allude to its meaning but it isn't obvious.

**Comment by quoo** [comment]
I always assumed it was short for queue, which makes sense.

**Comment by reddec** [comment]
I think that reason is just  42. `//IMHO`

**Comment by kriskowal** [comment]
The name Q came from a proposal to TC39 by Mark Miller (@erights) which in turn came from a library by Tyler Close in Waterken with similar shape and purpose. I have the impression that it is short for “event queue” since promises manage events, and the brevity is deliberate, to encourage folks to wrap promises they receive from others with `Q(promise).then...` instead of trusting `promise.then...` directly. Not all promises are made alike, but most can be tamed by virtue of thenability.

**Comment by erights** [comment]
Yes. My old TC39 draft proposal: http://wiki.ecmascript.org/doku.php?id=strawman:concurrency
Tyler's old Waterken Q API: http://waterken.sourceforge.net/web_send/#API

And yes, Q is mnemonic for "queue" used as a verb. Most of the API queues things.

**Comment by erights** [comment]
In looking for those, I also came across https://github.com/kriskowal/q/wiki/Coming-from-Waterken%E2%80%99s-Ref-Send which is interesting.

**Comment by kriskowal** [comment]
@erights That has not aged well. The main points about the limitations of using closures as objects to placate ADSafe became unnecessary in the world of ES5 and SES, so I altered the interface accordingly. The rest has varied in the details. Q no longer does valueOf. The treatment on variadic arguments varies by version as well. Version 2 eschews variadic arguments entirely for the internal handlers, grasping for performance.
